### PR TITLE
bundle plugin added with embedded mandatory dependencies

### PR DESCRIPTION
**Pull request by balazsjonas**
The goal of this pull request to make easier to use rtee in osgi environments.

Example can be find here:
https://github.com/balazsjonas/RtreeKarafTest

### CODE REVIEW

**Comment by balazsjonas** [comment]
> Thanks for the contribution, I haven't made bundles before so this is new territory for me.
> I ran mvn clean install and saw some warnings:
> [INFO] --- maven-bundle-plugin:4.2.1:bundle (default-bundle) @ rtree ---
> [WARNING] Bundle com.github.davidmoten:rtree:bundle:0.8.8-SNAPSHOT : Export com.github.davidmoten.rtree,  has 4,  private references [com.github.davidmoten.guavamini, com.github.davidmoten.rtree.internal, rx, rx.functions]
> [WARNING] Bundle com.github.davidmoten:rtree:bundle:0.8.8-SNAPSHOT : Export com.github.davidmoten.rtree.geometry,  has 1,  private references [rx.functions]
> [WARNING] Bundle com.github.davidmoten:rtree:bundle:0.8.8-SNAPSHOT : Export com.github.davidmoten.rtree.fbs,  has 1,  private references [rx.functions]
> [WARNING] Bundle com.github.davidmoten:rtree:bundle:0.8.8-SNAPSHOT : Export com.github.davidmoten.rtree.kryo,  has 1,  private references [rx.functions]
> 
> What should we do about them?

I made a huge mistake. An invalid tag left in the code instead of Export-Package.

About your question. I think the first part of the warning is valid and I committed the correction. But I have to check this. 
In the example some classes from rtree.geometry were used.

The second part of the warning guava-mini and rxjava is embeded which means these will be available on classpath at [runtime](https://felix.apache.org/documentation/subprojects/apache-felix-maven-bundle-plugin-bnd.html#embedding-dependencies).

The third part is about optional dependencies. I think the user of the library should provide these.

File: pom.xml [hunk 1]
```diff
@@ -371,7 +371,11 @@
                 <extensions>true</extensions>
                 <configuration>
                     <instructions>
-                        <i-Package>com.github.davidmoten.rtree.*</i-Package>
+                        <Export-Package>
+                            com.github.davidmoten.rtree;
+                            com.github.davidmoten.rtree.geometry;
+                            com.github.davidmoten.rtree.internal;
```
Comment (davidmoten) [inline pom.xml:377]: Why would we export an internal package?

Comment (davidmoten) [reply pom.xml:377]: And why aren't we exporting all public packages?

Comment (balazsjonas) [reply pom.xml:377]: 1. Some classes have reference to internal classes, which should also be available to the users. Eg. Geometries refers to classes in rtree.geometry.internal.
2. The easiest way seems to export all public packages. This is the least invasive. I think we can use wildcard '*'.

Comment (davidmoten) [reply pom.xml:377]: I looked at RxJava bundle and it looks like this: 
```
'Export-Package': '!io.reactivex.rxjava3.internal.*, io.reactivex.rxjava3.*'
```
I assume that !  means don't export the internal stuff. We should do the same.

Comment (balazsjonas) [reply pom.xml:377]: I think something like this should work: 
!com.github.davidmoten.rtree.internal.*;
!com.github.davidmoten.rtree.geometry.*;
com.github.davidmoten.rtree.*;

Also [guava](https://github.com/google/guava/blob/master/guava/pom.xml) does the same.

File: pom.xml [hunk 2]
```diff
@@ -364,6 +364,31 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <version>4.2.1</version>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Export-Package>
+                            com.github.davidmoten.rtree;
+                            com.github.davidmoten.rtree.geometry;
+                            com.github.davidmoten.rtree.internal;
```
Comment (balazsjonas) [inline pom.xml:377]:
```suggestion
                            com.github.davidmoten.rtree;
                            com.github.davidmoten.rtree.geometry;
                            com.github.davidmoten.rtree.internal;
```

Comment (davidmoten) [reply pom.xml:377]: I still think it should be 

```
!com.github.davidmoten.rtree.internal.*;
com.github.davidmoten.rtree.*;
```
This matches the default behaviour of Felix plugin. From https://felix.apache.org/documentation/subprojects/apache-felix-maven-bundle-plugin-bnd.html#default-behavior:

>&lt;Export-Package&gt; is now assumed to be the set of packages in your local Java sources, excluding the default package '.' and any packages containing 'impl' or 'internal'. (before version 2 of the bundleplugin it was based on the symbolic name)

The internal package classes will still be available at runtime for use by the exported classes (but presumably not available in an OSGi context for direct instantiation). 

If you think any classes have been labelled internal yet they are exposed in the public API then let me know and I'll have a look with a view to moving them around.

Comment (balazsjonas) [reply pom.xml:377]: I think I'm convinced. I've extended the example with a usage of a class which has a reference to an internal class. Eg. when ```Entries.entry(object, geometry)``` is called, an instance of EntryDefault created. Both bundles may run on the same jvm only the visibility is restricted.

Comment (davidmoten) [reply pom.xml:377]: Thanks for trying it out. So I assume that you can't call `EntryDefault.entry(...)` for instance?

Comment (balazsjonas) [reply pom.xml:377]: Yes, I can't. The example bundle is not activated due to missing requirement.

**Comment by davidmoten** [comment]
resolved in [0.9](https://github.com/davidmoten/rtree/releases/tag/0.9) on Maven Central now

**Pull request merged**